apps/andi/lib/andi_web/input_schemas/metadata.ex

```diff
@@ -0,0 +1,102 @@
+defmodule AndiWeb.InputSchemas.Metadata do
+  @moduledoc false
+  import Ecto.Changeset
+
+  @business_fields %{
+    contactEmail: :string,
+    contactName: :string,
+    dataTitle: :string,
+    description: :string,
+    homepage: :string,
+    issuedDate: :date,
+    keywords: {:array, :string},
+    language: :string,
+    license: :string,
+    modifiedDate: :date,
+    orgTitle: :string,
+    publishFrequency: :string,
+    spatial: :string,
+    temporal: :string
+  }
+
+  @technical_fields %{
+    private: :boolean,
+    sourceFormat: :string
+  }
+
+  @types Map.merge(@business_fields, @technical_fields)
+
+  @email_regex ~r/^[A-Za-z0-9._%+-+']+@[A-Za-z0-9.-]+\.[A-Za-z]+$/
+
+  def changeset_from_struct(%SmartCity.Dataset{} = dataset) do
+    from_business = get_business(dataset.business)
+    from_technical = get_technical(dataset.technical)
+
+    Map.merge(from_business, from_technical) |> new_changeset()
+  end
+
+  def restruct(schema, dataset) do
+    formatted_schema =
+      schema
+      |> Map.update!(:issuedDate, &date_to_iso8601_datetime/1)
+      |> Map.update!(:modifiedDate, &date_to_iso8601_datetime/1)
+
+    business = Map.merge(dataset.business, get_business(formatted_schema))
+    technical = Map.merge(dataset.technical, get_technical(formatted_schema))
+
+    dataset
+    |> Map.put(:business, business)
+    |> Map.put(:technical, technical)
+  end
+
+  def new_changeset(params \\ %{}), do: changeset({%{}, @types}, params)
+
+  def changeset(changeset, params \\ %{}) do
+    params =
+      case params do
```

Comment:
I would stick this case statement in `update_keywords_params` to keep this function at a higher level of abstraction.

Reply:
We refactored this. We kept that in the changeset because the concern felt like it fell under the changesets responsibility to validate and transform raw input data.

But we split it to be more clearly separate the logic for handling a list of keywords from the struct from the logic to handle the raw string from the form.